libsolidity/codegen/ir/IRVariable.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libsolidity/codegen/ir/IRVariable.h>
+#include <libsolidity/ast/AST.h>
+#include <boost/range/adaptor/transformed.hpp>
+#include <libsolutil/StringUtils.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::frontend;
+using namespace solidity::util;
+
+IRVariable::IRVariable(VariableDeclaration const& _declaration):
+	m_type(*_declaration.annotation().type),
+	m_baseName("vloc_" + _declaration.name() + '_' + std::to_string(_declaration.id()))
+{
+	solAssert(!_declaration.isStateVariable(), "");
+}
+
+IRVariable::IRVariable(Type const& _type, std::string _name):
+	m_type(_type), m_baseName(std::move(_name))
+{
+}
+
+IRVariable::IRVariable(Expression const& _expression):
+	m_type(*_expression.annotation().type),
+	m_baseName("expr_" + to_string(_expression.id()))
+{
+}
+
+IRVariable IRVariable::part(std::string const& _slot, Type const& _type) const
+{
+	return IRVariable(_type, part(_slot));
```

Comment:
Is this the right constructor call?

Reply:
Why wouldn't it be? Maybe the discrepency in the argument order is unfortunate... but apart from that it is what I want: no conversion, just consider the stack slot to have a specific type...

It'd be nicer not to have this function at all, but then I'd need to actually type the slots in the type system...

Consider ``_address`` for functions. This member function here exists, so that correct type conversion happens for
``define(var.part("_address"), ...)``... because currently ``var`` has no way of knowing that this slot has an address type, so I have to tell it explicitly...
We could make the stack slot lists in the types themselves be typed to solve this, but I'm not sure about that...

Reply:
Maybe typing the slot is actually not the worst of ideas... I almost went for it already earlier... let me know what you think - if we do, then this whole function vanishes, resp. merges with the one below.